### PR TITLE
ISPN-12319 XSite Response should allow a value

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-12319

### CODE REVIEW

File: core/src/main/java/org/infinispan/remoting/transport/RetryOnFailureXSiteCommand.java [hunk 1]
```diff
@@ -18,16 +18,16 @@
...
    public static final RetryPolicy NO_RETRY = new MaxRetriesPolicy(0);
    private static final Log log = LogFactory.getLog(RetryOnFailureXSiteCommand.class);
    private static final boolean trace = log.isTraceEnabled();
    private final XSiteBackup xSiteBackup;
-   private final XSiteReplicateCommand command;
+   private final XSiteReplicateCommand<O> command;
    private final RetryPolicy retryPolicy;
 
-   private RetryOnFailureXSiteCommand(XSiteBackup backup, XSiteReplicateCommand command, RetryPolicy retryPolicy) {
+   private RetryOnFailureXSiteCommand(XSiteBackup backup, XSiteReplicateCommand<O> command, RetryPolicy retryPolicy) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/remoting/transport/RetryOnFailureXSiteCommand.java:30]: can you add the generics to `org.infinispan.remoting.transport.RetryOnFailureXSiteCommand#newInstance()`?
just to remove the warnings from IDEA

File: core/src/main/java/org/infinispan/remoting/transport/impl/XSiteResponseImpl.java [hunk 2]
```diff
@@ -38,12 +39,12 @@ public void whenCompleted(XSiteResponseCompleted xSiteResponseCompleted) {
...
 
    @Override
-   public void accept(Object o, Throwable throwable) {
+   public void accept(ValidResponse response, Throwable throwable) {
       durationNanos = timeService.timeDuration(sendTimeNanos, TimeUnit.NANOSECONDS);
       if (throwable != null) {
          completeExceptionally(throwable);
       } else {
-         complete(null);
+         complete((O) response.getResponseValue());
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/remoting/transport/impl/XSiteResponseImpl.java:47]: nitpick
```suggestion
         //noinspection unchecked
         complete((O) response.getResponseValue());
```

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java [hunk 3]
```diff
@@ -200,13 +202,20 @@ private static PrivateMetadata internalMetadata(IracMetadata metadata) {
...
       try {
-         //currently, it only handles sync xsite requests.
-         //async xsite requests are handle by the other methods.
-         assert !preserveOrder;
-         //noinspection unchecked
-         return (CompletableFuture<Void>) command.acceptVisitor(null, defaultHandler);
+         if (command instanceof VisitableCommand) {
+            //noinspection unchecked
+            return (CompletionStage<Object>) ((VisitableCommand)command).acceptVisitor(null, defaultHandler);
+         } else if (command instanceof CacheRpcCommand) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:213]: -1 `CacheRpcCommand` doesn't require anything from the `BackupReceiver` and you have the `ComponentRegistry` available in `GlobalInboundInvocationHandler`

I would change `XSiteReplicateCommand` to the following and make the if-else in `SingleXSiteRpcCommand`
```
   public CompletionStage<O> performInLocalSite(ComponentRegistry registry, boolean preserveOrder) {
      //by default, the command is executed against BackupReceiver.
      return execute(registry.getBackupReceiver().running(), preserveOrder);
   }
   //abstract? or not?
   public abstract CompletionStage<O> execute(BackupReceiver receiver, boolean preserveOrder);
```

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java [hunk 4]
```diff
@@ -200,13 +202,20 @@ private static PrivateMetadata internalMetadata(IracMetadata metadata) {
...
+            //noinspection unchecked
+            return (CompletionStage<Object>) ((CacheRpcCommand) command).invokeAsync(componentRegistry);
+         } else {
+            throw new UnsupportedOperationException("Only VisitableCommand and CacheRpcCommand is allowed!");
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:217]: Also, don't need to spend CPU throwing the exception. you can create the CF with it directly (`completedExceptionFuture()`)

**Review by pruivo — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/infinispan/remoting/transport/RetryOnFailureXSiteCommand.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/remoting/transport/RetryOnFailureXSiteCommand.java:30]: Sure, just missed that one.

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java (hunk 4 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:217]: I personally don't think it is worth to add more code for a case that shouldn't ever happen.

File: core/src/main/java/org/infinispan/remoting/transport/impl/XSiteResponseImpl.java (hunk 2 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/remoting/transport/impl/XSiteResponseImpl.java:47]: 👍

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:213]: Okay, I will undo the ClusteredCacheBackupReceiver part and change it up a bit more.

**Comment by wburns** [comment]
@pruivo Should be all fixed up now, thanks!

File: core/src/main/java/org/infinispan/xsite/SingleXSiteRpcCommand.java [hunk 5]
```diff
@@ -34,8 +37,23 @@ public SingleXSiteRpcCommand() {
...
 
    @Override
-   public CompletionStage<Void> performInLocalSite(BackupReceiver receiver, boolean preserveOrder) {
-      return receiver.handleRemoteCommand(command, preserveOrder);
+   public CompletionStage<Object> performInLocalSite(ComponentRegistry registry, boolean preserveOrder) {
+      // Need to check VisitableCommand before CacheRpcCommand as PrepareCommand implements both but need to visit
+      if (command instanceof VisitableCommand) {
+         return super.performInLocalSite(registry, preserveOrder);
+      }
+      else {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/SingleXSiteRpcCommand.java:45]: nitpick:
```suggestion
      } else {
```

File: core/src/main/java/org/infinispan/xsite/SingleXSiteRpcCommand.java [hunk 6]
```diff
@@ -34,8 +37,23 @@ public SingleXSiteRpcCommand() {
...
+         try {
+            return (CompletionStage<Object>) ((CacheRpcCommand) command).invokeAsync(registry);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/SingleXSiteRpcCommand.java:47]: nitpick:
```suggestion
            //noinspection unchecked
            return (CompletionStage<Object>) ((CacheRpcCommand) command).invokeAsync(registry);
```

File: core/src/test/java/org/infinispan/jmx/PerCacheInboundHandlerMBeanTest.java [hunk 7]
```diff
@@ -82,7 +82,7 @@ public void testStats() throws Throwable {
       assertEquals(Boolean.TRUE, mBeanServer.getAttribute(objName, "StatisticsEnabled"));
 
       XSiteReplicateCommand command = mock(XSiteReplicateCommand.class);
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/jmx/PerCacheInboundHandlerMBeanTest.java:84]: nitpick:
```suggestion
      XSiteReplicateCommand<?> command = mock(XSiteReplicateCommand.class);
```

**Review by pruivo — COMMENTED** [review]
just a minor thing.

**Comment by wburns** [comment]
@pruivo Made changes as suggested.

**Comment by pruivo** [comment]
integrated! thanks @wburns !

**Pull request merged**